### ISSUE TITLE
wasm-ld returns duplicate symbol errors while native gcc toolchain works fine

### ISSUE THREAD

**Issue by vadimkantorov** [emscripten-core/emscripten#14973]
It seems that I indeed do have duplicate symbols, but gcc resolves it somehow. Not sure if gcc is being correct though...

**Comment by sbc100** [comment]
Can you give an example?    Are the inputs being passed to gcc and emcc both the same form (e.g. object files and ar archives) or are you using something like LTO or bitcode linking?

**Comment by vadimkantorov** [comment]
It's hard to give a small example, since it's the last step of my TexLive building pipeline that takes > 1h to build.

I think that somehow gcc's ld decides to allow multiple definitions (though I have a suspicion that the code in the multiple defined symbols is actually not identical) and somehow does it by default, while wasm-ld doesn't seem to support this mode.

**Comment by sbc100** [comment]
The only difference that we would expect to see relates to `common` symbols.   The llvm WebAssembly backend doesn't support common symbols and defaults to `-fno-common`.   Does adding `-fno-common` to your gcc build cause it to display the same bahaviour?

Do the symbols in question look like common symbols: https://gcc.gnu.org/onlinedocs/gcc/Code-Gen-Options.html

```
-fcommon
In C code, this option controls the placement of global variables defined without an initializer, known as tentative definitions in the C standard. Tentative definitions are distinct from declarations of a variable with the extern keyword, which do not allocate storage.

The default is -fno-common, which specifies that the compiler places uninitialized global variables in the BSS section of the object file. This inhibits the merging of tentative definitions by the linker so you get a multiple-definition error if the same variable is accidentally defined in more than one compilation unit.

The -fcommon places uninitialized global variables in a common block. This allows the linker to resolve all tentative definitions of the same variable in different compilation units to the same object, or to a non-tentative definition. This behavior is inconsistent with C++, and on many targets implies a speed and code size penalty on global variable references. It is mainly useful to enable legacy code to link without errors.
```

If this is not the issue then you might want to try building with `-Wl,--trace-symbol=xxxx` with both the gcc and the emcc build to see why one contains duplicate symbols but not the other?

Are you passing anything like `-Wl,--whole-archive` to emcc?  Or building with `MAIN_MODULE=1`?  Or doing bitcode linking?   Any of those things could also cause duplicate symbols.

**Comment by vadimkantorov** [comment]
I'll try g++ with `-fno-common` and will let you know

```
$(CXX_native) -Wl,--unresolved-symbols=ignore-all -Wimplicit -Wreturn-type -export-dynamic $(CFLAGS_OPT_native) -o $@ $@.o $(addprefix build/native/texlive/texk/web2c/, $(OBJ_XETEX) $(OBJ_PDFTEX)) $(addprefix build/native/, $(OBJ_BIBTEX) $(OBJ_DVIPDF) $(OBJ_DEPS)) $(addprefix -Ibuild/native/, $(CPATH_BUSYTEX)) $(addprefix build/native/texlive/texk/kpathsea/, $(OBJ_KPATHSEA)) -ldl -lm -pthread

$(CXX_wasm) -Wl,-error-limit=0 -Wl,--unresolved-symbols=ignore-all $(CFLAGS_OPT) -o $@ $(basename $@).o -export-dynamic -s TOTAL_MEMORY=$(TOTAL_MEMORY) -s EXIT_RUNTIME=0 -s INVOKE_RUN=0  -s ASSERTIONS=1 -s ERROR_ON_UNDEFINED_SYMBOLS=0 -s FORCE_FILESYSTEM=1 -s LZ4=1 -s MODULARIZE=1 -s EXPORT_NAME=$(notdir $(basename $@)) -s EXPORTED_FUNCTIONS='["_main", "_fflush", "_putchar", "_fopen", "_fputc", "_flush_streams"]' -sDEFAULT_LIBRARY_FUNCS_TO_INCLUDE=__sys_statfs64 -s EXPORTED_RUNTIME_METHODS='["callMain","FS", "ENV", "allocateUTF8OnStack", "LZ4", "PATH"]'  -lm $(addprefix build/wasm/texlive/texk/web2c/, $(OBJ_XETEX) $(OBJ_PDFTEX) ) $(addprefix build/wasm/, $(OBJ_DVIPDF) $(OBJ_BIBTEX) $(OBJ_DEPS)) $(addprefix -Ibuild/wasm/, $(CPATH_BUSYTEX))
```

**Comment by vadimkantorov** [comment]
Adding `-fno-common` to g++ build line did not break it.

**Comment by vadimkantorov** [comment]
@sbc100 Ok, I think I have a hunch: they are all static variables, not functions

**Comment by sbc100** [comment]
Can you point me to the declaration of one of these duplicate symbols.  It does seem it could well be common symbols that use of common symbols.

Lets take one symbol for example:

```
wasm-ld: error: duplicate symbol: insertsrcspecialeveryvbox                      
  >>> defined in build/wasm/texlive/texk/web2c/xetexdir/xetex-xetexextra.o         
  >>> defined in build/wasm/texlive/texk/web2c/pdftexdir/pdftex-pdftexextra.o  
```

is this essentially the same object file being included twice?   It makes sense that one would see duplicate symbols if one tried to include the same object trice on the command line.   One way to avoid such errors would be to using `ar` archives where entire objects can be ignored if they are not needed (presumably in this case you really only one of these two different object files to be selected by the linker?)

I still don't see of how the `gcc -fno-common` would not also display duplicate symbol errors?  Are both those object files listed directly on the command line (and not in ar archives?) in the gcc build too?

**Comment by vadimkantorov** [comment]
Yes, both these objects are included and both of them have functions / symbols that are not identical. But they also contain identical symbols such as these static variables.

Yes, both object files are always listed directly on the command line (including gcc build). I added a trace switch for one of these duplicate symbols.

**Comment by vadimkantorov** [comment]
E.g. definition of array `write_file`: https://github.com/TeX-Live/texlive-source/blob/trunk/texk/web2c/tex.web#L24604

**Comment by sbc100** [comment]
I have no idea how to read that :)   That does file produce C code as output somehow?

If variables are declared as `static` they should never be reported as duplicate.  Only global variables should ever be reported as conflicting/duplicate between object files.   Can you confirm that they are static?   The linker certainly doesn't think they are.

**Comment by vadimkantorov** [comment]
The converted files are: 
[pdftexd.h](https://github.com/emscripten-core/emscripten/files/7090273/pdftexd.h.txt)
[xetexd.h](https://github.com/emscripten-core/emscripten/files/7090274/xetexd.h.txt)

**Comment by vadimkantorov** [comment]
Yes, WEB is a dialect of Pascal invented by Knuth :)

**Comment by sbc100** [comment]
Looking at the declaration of `insertsrcspecialeverymath` for example I see:

`EXTERN boolean insertsrcspecialeverymath  ;`

So it depends what `EXTERN` is defined as but my guess its that this is declaring a COMMON data symbol.   

Can you find a the definition of `EXTERN` ?

**Comment by vadimkantorov** [comment]
`xetexini.c:#define EXTERN extern`

**Comment by vadimkantorov** [comment]
It would seem that if the data-sizes for static variables are matching, gcc would succeed in linking

Still not sure why it doesn't break for all other duplicate symbols in *.o files...

**Comment by sbc100** [comment]
I can only find declarations of `EXTERN boolean writeopen[18]` in those files you attached and never a definition.   Is there a place where `writeopen` is declared static?  Or is `EXTERN` perhaps sometimes defined to be `static` rather than `extern?

**Comment by sbc100** [comment]
This is exactly what `-fcommon` is for, which make me think that your gcc build must be using `-fcommon`.

```
$ cat test1.c 
int writeopen[18];
$ cat test2.c 
int writeopen[18];
$ gcc -fcommon test2.c test1.c main.c
$ gcc -fno-common test2.c test1.c main.c
/usr/bin/ld: /tmp/ccVP1U6l.o:(.bss+0x0): multiple definition of `writeopen'; /tmp/ccjhGyXk.o:(.bss+0x0): first defined here
```

**Comment by vadimkantorov** [comment]
```shell
echo "extern int writeopen[18];" > test1.c
echo "extern int writeopen[18];" > test2.c
echo "main() {}" > main.c

# works
gcc -fno-common test2.c test1.c main.c
```

**Comment by vadimkantorov** [comment]
not if they are defined as extern

**Comment by vadimkantorov** [comment]
i checked, they are only defined as extern, never as static

**Comment by sbc100** [comment]
In they are only ever `extern` the would show up as undefined .. somebody needs to define them.

Add a reference to writeopen to your main function:

```
$ cat main.c
extern int writeopen[18];
#include <stdio.h>

int main() {
  printf("%d\n", writeopen[0]);
}
```

**Comment by sbc100** [comment]
`extern` does not produce a definition so they must be defined as something other than `extern` in both the object files.

**Comment by vadimkantorov** [comment]
I'm now running a build to dump all object files as Github actions artifact. I'll put them here once I have them!

Thanks for help!

**Comment by vadimkantorov** [comment]
Here https://www.tutorialspoint.com/extern-keyword-in-c they say that initialization of extern variable is considered as implicit definition

**Comment by vadimkantorov** [comment]
Found it. In https://github.com/TeX-Live/texlive-source/blob/trunk/texk/web2c/pdftexdir/pdftexextra.c#L5

and the corresponding XeTeX file EXTERN is reset to empty modifier, and hence is static?

**Comment by sbc100** [comment]
The empty modifier is not static its global.    Unless you have `-fcommon` that would result in duplicate definition.

**Comment by vadimkantorov** [comment]
I think I made a mistake, looked at the wrong logs. Now re-running "-fno-common" gcc build

**Comment by sbc100** [comment]
`-fno-common` is also now the default for gcc 10 and above apparently (https://gcc.gnu.org/bugzilla/show_bug.cgi?id=85678).  So on my machine you need to opt into this with `-fcommon` 

Opting into `-fcommon` will fail under emcc:

```
$ emcc -fcommon test1.c -c
Common symbols are not yet implemented for Wasm
UNREACHABLE executed at /b/s/w/ir/cache/builder/emscripten-releases/llvm-project/llvm/lib/MC/MCWasmStreamer.cpp:139!
```

**Comment by vadimkantorov** [comment]
Hmm, I've checked, the `-fno-common`, indeed succeeded:
```
g++ -Wl,--unresolved-symbols=ignore-all -fno-common -Wimplicit -Wreturn-type -export-dynamic -O3 -o build/native/busytex build/native/busytex.o build/native/texlive/texk/web2c/synctexdir/xetex-synctex.o build/native/texlive/texk/web2c/xetex-xetexini.o build/native/texlive/texk/web2c/xetex-xetex0.o build/native/texlive/texk/web2c/xetex-xetex-pool.o build/native/texlive/texk/web2c/xetexdir/xetex-xetexextra.o build/native/texlive/texk/web2c/lib/lib.a build/native/texlive/texk/web2c/libmd5.a build/native/texlive/texk/web2c/busytex_libxetex.a build/native/texlive/texk/web2c/synctexdir/pdftex-synctex.o build/native/texlive/texk/web2c/pdftex-pdftexini.o build/native/texlive/texk/web2c/pdftex-pdftex0.o build/native/texlive/texk/web2c/pdftex-pdftex-pool.o build/native/texlive/texk/web2c/pdftexdir/pdftex-pdftexextra.o build/native/texlive/texk/web2c/lib/lib.a build/native/texlive/texk/web2c/libmd5.a build/native/texlive/texk/web2c/busytex_libpdftex.a build/native/texlive/texk/bibtex-x/busytex_bibtex8.a build/native/texlive/texk/dvipdfm-x/busytex_xdvipdfmx.a build/native/texlive/libs/harfbuzz/libharfbuzz.a build/native/texlive/libs/graphite2/libgraphite2.a build/native/texlive/libs/teckit/libTECkit.a build/native/texlive/libs/libpng/libpng.a build/native/fontconfig/src/.libs/libfontconfig.a build/native/texlive/libs/freetype2/libfreetype.a build/native/texlive/libs/pplib/libpplib.a build/native/texlive/libs/zlib/libz.a build/native/texlive/libs/zziplib/libzzip.a build/native/texlive/libs/libpaper/libpaper.a build/native/texlive/libs/icu/icu-build/lib/libicuuc.a build/native/texlive/libs/icu/icu-build/lib/libicudata.a build/native/texlive/libs/lua53/.libs/libtexlua53.a build/native/texlive/libs/xpdf/libxpdf.a build/native/texlive/texk/kpathsea/.libs/libkpathsea.a build/native/expat/libexpat.a -Ibuild/native/texlive/libs/icu/include -Ibuild/native/fontconfig build/native/texlive/texk/kpathsea/busytex_kpsewhich.o build/native/texlive/texk/kpathsea/.libs/libkpathsea.a -ldl -lm -pthread 
```

**Comment by vadimkantorov** [comment]
I didn't quite understand, is there a common/no-common mode I should try with g++?

`-fno-common` compiled fine

**Comment by vadimkantorov** [comment]
Maybe somehow `g++` on Github Actions worker ignored `-fno-common`?

**Comment by sbc100** [comment]
I think you need to compile (not link) with `-fno-common` to get the emcc-link behaviour.  Do you know what gcc version you are you using?

**Comment by vadimkantorov** [comment]
Ah I see

**Comment by vadimkantorov** [comment]
I'm not sure about the version, it's default Ubuntu 20.04 worker on Github Actions, I'll check

**Comment by sbc100** [comment]
https://packages.ubuntu.com/focal/gcc looks like 9.3.0 so `-fcommon` will be the default which is why it working.

**Comment by vadimkantorov** [comment]
I guess the root cause is roughly uncovered, so I renamed the issue accordingly

**Comment by vadimkantorov** [comment]
Is there a way I could redefine EXTERN to add a modifier that would make them hidden in one of object files?

**Comment by sbc100** [comment]
Yes you could just defined it as `static`.   It really depends if its intent is to actually share the variables between the different files of it its OK for each file to have its own `writeopen`.  I suspect the sharing is indended..   in which case the correct fix would be to choose a central location in which to define them and have all the other files use `extern`.

**Comment by thegreatfatzby** [comment]
Hey @sbc100 commenting on this here as it's emscripten specific part of my issue (see https://github.com/llvm/llvm-project/issues/55270#issuecomment-2216120346) and this one is open...I do see emcc pass on the -fcommon to the downstream clang command it constructs, if putting a hard error in clang (like -fcommon is incompatible with wasm, please try again) isn't a good idea, maybe we could put something in emcc command, or possibly even just the [docs](https://emscripten.org/search.html?q=fcommon&check_keywords=yes&area=default) indicating that this won't work?

**Comment by sbc100** [comment]
I don't think this is really and emscripten issue.  We should focus on the upstream llvm issue.  I think having a better error message when common symbols are encountered should be fine (if you can figure out how to do it).
